### DISCUSSION TITLE
struct/enum fields names

### DISCUSSION THREAD

**Discussion by sdancer** [gleam-lang/gleam#1748] [Ideas & suggestions]
can't mix underscores or other symbols to names, how to define a large protocol ?

imagine a few hundreds of these used on real code to implement a network protocol:
```
pub type OptionType {
  OptionTypeFoo
  OptionTypeBar
  OptionTypeBaz
}
pub type ExampleResponseOrError {
  ExampleResponseOrErrorResponse(response: option.Option(Response))
  ExampleResponseOrErrorError(error: String)
}
```
in other langs, u at least can use some symbols on the names
```
pub type OptionType {
  OptionType_Foo
  OptionType_Bar
  OptionType_Baz
}
```
in rust u have multiple alternatives to avoid name clashes, and **reduce visual clutter** or at least hint where a name ends and another stars

**Comment by sporto** [comment]
This is something I do often in Elm. Being able to use underscores in names is very useful to make the namespaces of custom types clear.

However, would be even better being able to write this like Rust e.g. `ExampleResponseOrError::Response`.

**Comment by sdancer** [comment]
@sporto of course, rust-like would be ideal, but i guess it might require more complex rework on the compiler

**Comment by lpil** [comment]
It is likely too late in the development in the language to make a change as large as making custom type constructors qualified, so we would need to come up with an appealling backwards compatible design or one with a very good upgrade path.

We could permit underscores in constructor names but we would need to determine how to use these in code generation. We currently convert names to snake case to use in Erlang, we would need to continue to do this but also ensure that no two names clash after conversion. i.e. `OneTwo` does not render the same as `One_Two` or `One__Two` or `OneTwo_`

**Reply by timjs** [reply]
Isn't or wouldn't it be possible to refer to an constructor using the module name it came from? I.e. `result.Ok`

**Reply by lpil** [reply]
Yes, this is the recommended approach today. I'm quite happy with it but it doesn't solve the problem if you wish to have multiple constructors with the same name in a single module. Perhaps we could solve this problem but for me it is niche enough to not warrant much effort currently.

**Reply by timjs** [reply]
Ah yeah. If some would like constructors of the same name, then using another module is the right thing to do I think.
